include/stl2/detail/to.hpp

```diff
@@ -0,0 +1,168 @@
+			return __class_to<_Cont>{}(static_cast<_IRange&&>(__r));
+		}
+
+		template<class _Cont, class _First, class... _Rest>
+		requires __class_convertible<_Cont, _First, _Rest...> ||
+			(CopyConstructible<_First> && (CopyConstructible<_Rest> && ...))
+		constexpr auto to(_First&& __first, _Rest&&... __rest) {
+			if constexpr (__class_convertible<_Cont, _First, _Rest...>) {
+				return __class_to<_Cont>{}(
+					static_cast<_First&&>(__first), static_cast<_Rest&&>(__rest)...);
+			} else {
+				return __class_closure<_Cont, _First, _Rest...>{
+					static_cast<_First&&>(__first), static_cast<_Rest&&>(__rest)...};
+			}
+		}
+
+		template<template<class...> class _ContT, InputRange _IRange, class... _Args>
+		using __ctad = decltype(_ContT(__range_common_iterator<_IRange>(),
+			__range_common_iterator<_IRange>(), std::declval<_Args>()...));
+
+		template<template<class...> class _ContT, class _IRange, class... _Args>
+		META_CONCEPT __CTADConvertible = InputRange<_IRange> && requires {
+			typename __ctad<_ContT, _IRange, _Args...>;
+			requires __class_convertible<__ctad<_ContT, _IRange, _Args...>, _IRange, _Args...>;
+		};
+
+		template<template<class...> class _ContT, class _IRange, class... _Args>
+		META_CONCEPT __NoCTADConvertible = InputRange<_IRange> && requires {
+			typename _ContT<ext::range_value_t<_IRange>>;
+			requires __class_convertible<_ContT<ext::range_value_t<_IRange>>, _IRange, _Args...>;
+		};
+
+		template<template<class...> class _ContT, class _IRange, class... _Args>
+		META_CONCEPT __template_convertible = __CTADConvertible<_ContT, _IRange, _Args...> ||
+			__NoCTADConvertible<_ContT, _IRange, _Args...>;
+
+		template<template<class...> class _ContT>
+		struct __template_to {
+			template<class _IRange, class... _Args>
+			requires __template_convertible<_ContT, _IRange, _Args...>
+			constexpr auto operator()(_IRange&& __r, _Args&&... __args) const {
+				if constexpr (__CTADConvertible<_ContT, _IRange, _Args...>) {
+					return __class_to<__ctad<_ContT, _IRange, _Args...>>{}(
+						static_cast<_IRange&&>(__r), static_cast<_Args&&>(__args)...);
+				} else {
+					return __class_to<_ContT<ext::range_value_t<_IRange>>>{}(
+						static_cast<_IRange&&>(__r), static_cast<_Args&&>(__args)...);
+				}
+			}
+		};
+
+		template<template<class...> class _ContT, class... _Args>
+		requires (CopyConstructible<_Args> && ...)
+		struct STL2_EMPTY_BASES __template_closure
+		: detail::closure<__template_to<_ContT>, _Args...> {
+			constexpr __template_closure(_Args&&... __args)
+			: detail::closure<__template_to<_ContT>, _Args...>{
+				{}, static_cast<_Args&&>(__args)...} {}
+		};
+
+		template<template<class...> class _ContT>
+		constexpr __template_closure<_ContT> to() noexcept {
+			return {};
+		}
+
+		template<template<class...> class _ContT, class _IRange>
+		requires __template_convertible<_ContT, _IRange>
+		constexpr auto operator|(_IRange&& __r, __template_closure<_ContT>(&)() noexcept) {

```

Comment:
Same feedback here. I have it on good authority\[*\] that this doesn't work in general.

\[*\]: https://twitter.com/ericniebler/status/1085600729656549377